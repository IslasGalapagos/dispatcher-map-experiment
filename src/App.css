.map-holder {
  float: left;
  width: 80vw;
  height: 100vh;
}

.leaflet-container {
  width: 100%;
  height: 100%;
}

.list-holder {
  float: right;
  height: 100vh;
  width: 20vw;
  max-height: 100vh;
  overflow: scroll;
  overflow-x: hidden;
}

.list-holder--item {
  padding: 10px;
  border-bottom: 4px solid #444;
}

.list-holder--item:last-child {
  border-bottom: none;
}

.list-holder--item--name {
  font-weight: bold;
}

.draggable-holder {
  position: absolute;
  background-color: white;
  z-index: 1000;
  padding: 10px;
  width: 200px;
  border: 4px solid #444;
}

.draggable-holder--hide {
  position: absolute;
  width: 30px;
  height: 30px;
  border: 4px solid #444;
  border-radius: 50%;
  right: -18px;
  top: -18px;
  cursor: pointer;
}

.draggable-holder--hide::after {
  content: 'âœ•';
  position: absolute;
  width: 30px;
  height: 30px;
  top: 5px;
  left: -1px;
  font-weight: bolder;
  text-align: center;
}

.draggable-holder--name {
  font-weight: bold;
}

.work-order-icon--pin {
  width: 30px;
  height: 30px;
  border-radius: 50% 50% 50% 0;
  background: #89849b;
  position: absolute;
  transform: rotate(-45deg);
  left: 50%;
  top: 50%;
  margin: -20px 0 0 -20px;
  animation-name: bounce;
  animation-fill-mode: both;
  animation-duration: 1s;
  border: 4px solid #444;
}

.draggable-holder--DROPOFF-CAN,
.draggable-holder--DROPOFF-CAN .draggable-holder--hide,
.work-order-icon--DROPOFF-CAN .work-order-icon--pin,
.list-holder--item--DROPOFF-CAN {
  background: #e9d1ff;
}

.draggable-holder--FINAL,
.draggable-holder--FINAL .draggable-holder--hide,
.work-order-icon--FINAL .work-order-icon--pin,
.list-holder--item--FINAL {
  background: #fa3a37;
  color: white;
}

.draggable-holder--SPOT,
.draggable-holder--SPOT .draggable-holder--hide,
.work-order-icon--SPOT .work-order-icon--pin,
.list-holder--item--SPOT {
  background: #5fde6c;
}

.work-order-icon--pin .work-order-icon--inner {
  position: absolute;
  transform: rotate(45deg) translate(8px, 5px);
  width: 30px;
  height: 30px;
}

.work-order-icon--pulse {
  background: rgba(26,26,26,0.5);
  border-radius: 50%;
  height: 14px;
  width: 14px;
  position: absolute;
  left: 50%;
  top: 50%;
  margin: 19px 0px 0px -8px;
  transform: rotateX(55deg);
  z-index: -2;
}

.work-order-icon--pulse::after {
    content: "";
    border-radius: 50%;
    height: 40px;
    width: 40px;
    position: absolute;
    margin: -13px 0 0 -13px;
    animation: pulsate 1s ease-out;
    animation-iteration-count: infinite;
    opacity: 0.0;
    box-shadow: 0 0 1px 2px #89849b;
    animation-delay: 1.1s;
}

@keyframes pulsate {
  0% {
    transform: scale(0.1, 0.1);
    opacity: 0.0;
  }
  50% {
    opacity: 1.0;
  }
  100% {
    transform: scale(1.2, 1.2);
    opacity: 0;
  }
}

.drop-area {
  position: absolute;
  background-color: white;
  opacity: 0.5;
  top: 0;
  left: 0;
  width: 80vw;
  height: 100vh;
  z-index: 1001;
}
