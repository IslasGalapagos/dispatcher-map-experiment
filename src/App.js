import React, { Component } from "react";
import { connect } from "react-redux";
import { Map, TileLayer } from "react-leaflet";
import MarkerClusterGroup from "./third-party/MarkerClusterGroup";
import { bounds } from "./gen-points";
import "./App.css";

export const mapState = state => ({
  holder: state.holder,
  dropArea: state.dropArea,
  points: state.points || [],
  orders: state.orders
});

export const mapActions = dispatch => ({
  pointerOver: ({ _icon, options: { id, type, address } }) =>
    dispatch({
      type: "holder->show",
      rect: _icon.getBoundingClientRect(),
      data: { id, type, address }
    }),
  pointerOut: () =>
    dispatch({
      type: "holder->hide"
    }),

  moveToList: data => dispatch({ type: "list->move", data }),

  startMoveOutList: order => dispatch({ type: "list->moveStart", order }),
  stopMoveOutList: order => dispatch({ type: "list->moveStop" }),
  moveOutList: data => dispatch({ type: "list->moveOut", data })
});

export const prevent = event => event.preventDefault();

export class App extends Component {
  render() {
    const {
      moveToList,
      startMoveOutList,
      stopMoveOutList,
      moveOutList,
      pointerOver,
      pointerOut,
      pointerDragEnd,
      points,
      orders,
      holder,
      dropArea
    } = this.props;
    return (
      <div>
        <div className="map-holder">
          <Map bounds={bounds} maxZoom={18}>
            <TileLayer url="http://{s}.tile.osm.org/{z}/{x}/{y}.png" />
            <MarkerClusterGroup
              onMarkerMouseOver={pointerOver}
              markers={points}
            />
          </Map>
        </div>
        {holder &&
          <div
            className="draggable-holder"
            draggable="true"
            onDragStart={event =>
              event.dataTransfer.setData(
                "text/plain",
                JSON.stringify(holder.data)
              )}
            onDragEnd={pointerOut}
            onMouseOut={pointerOut}
            style={{
              top: holder.top,
              left: holder.left,
              width: holder.width,
              height: holder.height
            }}
          />}
        <div
          className="list-holder"
          onDragEnter={prevent}
          onDragOver={prevent}
          onDrop={event => {
            const data = JSON.parse(event.dataTransfer.getData("text"));
            if (!data.id) {
              return;
            }
            moveToList(data);
          }}
        >
          {orders.map(order => (
            <div
              key={order.options.id}
              draggable="true"
              onDragStart={event => {
                event.dataTransfer.setData("text/plain", JSON.stringify(order));
                startMoveOutList(order);
              }}
              onDragEnd={stopMoveOutList}
            >
              {order.options.type}<br />{order.options.address}<hr />
            </div>
          ))}
        </div>
        {dropArea &&
          <div
            className="drop-area"
            onDragEnter={prevent}
            onDragOver={prevent}
            onDrop={event => {
              const data = JSON.parse(event.dataTransfer.getData("text"));
              moveOutList(data);
              stopMoveOutList();
            }}
          />}
      </div>
    );
  }
}

export default connect(mapState, mapActions)(App);
