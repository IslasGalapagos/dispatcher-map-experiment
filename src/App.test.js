import React from "react";
import { App, mapState, mapActions, prevent } from "./App";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

describe("App", () => {
  it("render points", () => {
    const getData = jest.fn();
    getData.mockReturnValueOnce("{}").mockReturnValueOnce('{"id":1}');
    const moveToList = jest.fn();

    const holder = shallow(
      <App points={[{ lat: 1, lng: 2 }]} orders={[]} moveToList={moveToList} />
    );
    const $holder = holder.find(".list-holder");

    expect(toJson(holder)).toMatchSnapshot();

    $holder.simulate("drop", { dataTransfer: { getData } });
    expect(getData).toBeCalled();
    expect(moveToList.mock.calls).toHaveLength(0);

    $holder.simulate("drop", { dataTransfer: { getData } });
    expect(moveToList).toBeCalledWith({ id: 1 });
  });

  it("render holder over the map", () => {
    const pointerOut = jest.fn();
    const setData = jest.fn();
    const pointerDragEnd = jest.fn();

    const holder = shallow(
      <App
        orders={[]}
        holder={{ top: 1, left: 2, width: 3, height: 4, data: { prop: 101 } }}
        pointerOut={pointerOut}
        pointerDragEnd={pointerDragEnd}
      />
    );
    const $holder = holder.find(".draggable-holder");

    expect(toJson(holder)).toMatchSnapshot();

    $holder.simulate("dragstart", { dataTransfer: { setData } });
    expect(setData).toBeCalledWith("text/plain", '{"prop":101}');

    $holder.simulate("dragend");
    expect(pointerOut).toBeCalled();
  });

  it("render drop area over the map", () => {
    const moveOutList = jest.fn();
    const stopMoveOutList = jest.fn();
    const getData = jest.fn();
    getData.mockReturnValue('{"prop": 100}');

    const holder = shallow(
      <App
        orders={[]}
        dropArea={true}
        moveOutList={moveOutList}
        stopMoveOutList={stopMoveOutList}
      />
    );
    const $area = holder.find(".drop-area");

    expect(toJson(holder)).toMatchSnapshot();

    $area.simulate("drop", { dataTransfer: { getData } });

    expect(getData).toBeCalledWith("text");
    expect(moveOutList).toBeCalledWith({ prop: 100 });
    expect(stopMoveOutList).toBeCalled();
  });

  it("render orders", () => {
    const order = { options: { id: 1, type: 2, address: 3 } };
    const startMoveOutList = jest.fn();
    const stopMoveOutList = jest.fn();
    const setData = jest.fn();

    const holder = shallow(
      <App
        orders={[order]}
        startMoveOutList={startMoveOutList}
        stopMoveOutList={stopMoveOutList}
      />
    );

    expect(toJson(holder)).toMatchSnapshot();
    const orderElement = holder.find(".list-holder > div");

    orderElement.simulate("dragstart", { dataTransfer: { setData } });
    expect(setData).toBeCalledWith("text/plain", JSON.stringify(order));
    expect(startMoveOutList).toBeCalledWith(order);

    orderElement.simulate("dragend");
    expect(stopMoveOutList).toBeCalled();
  });
});

test("prevent", () => {
  const preventDefault = jest.fn();
  prevent({ preventDefault });
  expect(preventDefault).toBeCalled();
});

test("mapState", () => {
  expect(mapState({ holder: 1, dropArea: 2, orders: 3 })).toMatchSnapshot();
  expect(mapState({ points: 1 })).toMatchSnapshot();
});

describe("mapActions", () => {
  let dispatch;
  beforeEach(() => dispatch = jest.fn());

  it("pointerOver", () => {
    const { pointerOver } = mapActions(dispatch);

    pointerOver({
      _icon: { getBoundingClientRect: () => "RECT" },
      options: { id: 1, size: 4, type: 2, address: 3 }
    });

    expect(dispatch.mock.calls[0]).toMatchSnapshot();
  });

  it("pointerOut", () => {
    const { pointerOut } = mapActions(dispatch);

    pointerOut();

    expect(dispatch.mock.calls[0]).toMatchSnapshot();
  });

  it("moveToList", () => {
    const { moveToList } = mapActions(dispatch);

    moveToList("data");

    expect(dispatch.mock.calls[0]).toMatchSnapshot();
  });

  it("startMoveOutList", () => {
    const { startMoveOutList } = mapActions(dispatch);

    startMoveOutList("order");

    expect(dispatch.mock.calls[0]).toMatchSnapshot();
  });

  it("stopMoveOutList", () => {
    const { stopMoveOutList } = mapActions(dispatch);

    stopMoveOutList();

    expect(dispatch.mock.calls[0]).toMatchSnapshot();
  });

  it("moveOutList", () => {
    const { moveOutList } = mapActions(dispatch);

    moveOutList("data");

    expect(dispatch.mock.calls[0]).toMatchSnapshot();
  });
});
