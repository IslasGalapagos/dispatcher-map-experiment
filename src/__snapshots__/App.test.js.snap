exports[`App render drop area over the map 1`] = `
<div>
  <div
    className="map-holder">
    <Map
      animate={false}
      bounds={
        Array [
          Array [
            51.51280224425956,
            -0.11681556701660155,
          ],
          Array [
            51.50211782162702,
            -0.14428138732910156,
          ],
        ]
      }
      maxZoom={18}
      useFlyTo={false}>
      <TileLayer
        url="http://{s}.tile.osm.org/{z}/{x}/{y}.png" />
    </Map>
  </div>
  <div
    className="list-holder"
    onDragOver={[Function]}
    onDrop={[Function]} />
  <div
    className="drop-area"
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]} />
</div>
`;

exports[`App render holder over the map 1`] = `
<div>
  <div
    className="map-holder">
    <Map
      animate={false}
      bounds={
        Array [
          Array [
            51.51280224425956,
            -0.11681556701660155,
          ],
          Array [
            51.50211782162702,
            -0.14428138732910156,
          ],
        ]
      }
      maxZoom={18}
      useFlyTo={false}>
      <TileLayer
        url="http://{s}.tile.osm.org/{z}/{x}/{y}.png" />
    </Map>
  </div>
  <div
    className="draggable-holder draggable-holder--undefined"
    draggable="true"
    onDragEnd={[Function]}
    onDragStart={[Function]}
    style={
      Object {
        "left": -16,
        "top": -17,
      }
    }>
    <div
      className="draggable-holder--hide"
      onClick={[Function]} />
    <div
      className="draggable-holder--name" />
    <div
      className="draggable-holder--size">
       YARDS
    </div>
    <div
      className="draggable-holder--address" />
  </div>
  <div
    className="list-holder"
    onDragOver={[Function]}
    onDrop={[Function]} />
</div>
`;

exports[`App render orders 1`] = `
<div>
  <div
    className="map-holder">
    <Map
      animate={false}
      bounds={
        Array [
          Array [
            51.51280224425956,
            -0.11681556701660155,
          ],
          Array [
            51.50211782162702,
            -0.14428138732910156,
          ],
        ]
      }
      maxZoom={18}
      useFlyTo={false}>
      <TileLayer
        url="http://{s}.tile.osm.org/{z}/{x}/{y}.png" />
      <Marker
        icon={
          NewClass {
            "_initHooksCalled": true,
            "options": Object {
              "className": "work-order-sorted-icon",
              "html": "<span>1</span>",
            },
          }
        }
        position={
          Array [
            undefined,
            undefined,
          ]
        } />
    </Map>
  </div>
  <div
    className="list-holder"
    onDragOver={[Function]}
    onDrop={[Function]}>
    <div
      className="list-holder--item list-holder--item--2"
      data-id={1}
      draggable="true"
      onDragEnd={[Function]}
      onDragStart={[Function]}>
      <div
        className="list-holder--item--name"
        data-id={1} />
      <div
        className="list-holder--item--size"
        data-id={1}>
         YARDS
      </div>
      <div
        className="list-holder--item--address"
        data-id={1}>
        3
      </div>
    </div>
  </div>
</div>
`;

exports[`App render points 1`] = `
<div>
  <div
    className="map-holder">
    <Map
      animate={false}
      bounds={
        Array [
          Array [
            51.51280224425956,
            -0.11681556701660155,
          ],
          Array [
            51.50211782162702,
            -0.14428138732910156,
          ],
        ]
      }
      maxZoom={18}
      useFlyTo={false}>
      <TileLayer
        url="http://{s}.tile.osm.org/{z}/{x}/{y}.png" />
      <MarkerClusterGroup
        markers={
          Array [
            Object {
              "lat": 1,
              "lng": 2,
            },
          ]
        }
        wrapperOptions={Object {}} />
      <Marker
        icon={
          NewClass {
            "_initHooksCalled": true,
            "options": Object {
              "className": "work-order-sorted-icon",
              "html": "<span>1</span>",
            },
          }
        }
        position={
          Array [
            3,
            4,
          ]
        } />
      <Marker
        icon={
          NewClass {
            "_initHooksCalled": true,
            "options": Object {
              "className": "work-order-sorted-icon",
              "html": "<span>2</span>",
            },
          }
        }
        position={
          Array [
            5,
            6,
          ]
        } />
      <Polyline
        color="#444"
        positions={
          Array [
            Array [
              3,
              4,
            ],
            Array [
              5,
              6,
            ],
          ]
        } />
    </Map>
  </div>
  <div
    className="list-holder"
    onDragOver={[Function]}
    onDrop={[Function]}>
    <div
      className="list-holder--item list-holder--item--undefined"
      data-id={10}
      draggable="true"
      onDragStart={[Function]}>
      <div
        className="list-holder--item--name"
        data-id={10} />
      <div
        className="list-holder--item--size"
        data-id={10}>
         YARDS
      </div>
      <div
        className="list-holder--item--address"
        data-id={10} />
    </div>
    <div
      className="list-holder--item list-holder--item--undefined"
      data-id={20}
      draggable="true"
      onDragStart={[Function]}>
      <div
        className="list-holder--item--name"
        data-id={20} />
      <div
        className="list-holder--item--size"
        data-id={20}>
         YARDS
      </div>
      <div
        className="list-holder--item--address"
        data-id={20} />
    </div>
  </div>
</div>
`;

exports[`mapActions moveOutList 1`] = `
Array [
  Object {
    "data": "data",
    "type": "list->moveOut",
  },
]
`;

exports[`mapActions moveToList 1`] = `
Array [
  Object {
    "data": "data",
    "type": "list->move",
  },
]
`;

exports[`mapActions pointerOut 1`] = `
Array [
  Object {
    "type": "holder->hide",
  },
]
`;

exports[`mapActions pointerOver 1`] = `
Array [
  Object {
    "data": Object {
      "address": 3,
      "id": 1,
      "name": undefined,
      "size": 4,
      "type": 2,
    },
    "rect": "RECT",
    "type": "holder->show",
  },
]
`;

exports[`mapActions reorder 1`] = `
Array [
  Object {
    "draggedId": 245,
    "targetId": 1000,
    "type": "list->reorder",
  },
]
`;

exports[`mapActions reorder 2`] = `
Array [
  Object {
    "draggedId": 345,
    "targetId": 2000,
    "type": "list->reorder",
  },
]
`;

exports[`mapActions reorder 3`] = `
Array [
  Object {
    "draggedId": NaN,
    "targetId": 3000,
    "type": "list->reorder",
  },
]
`;

exports[`mapActions startMoveOutList 1`] = `
Array [
  Object {
    "order": "order",
    "type": "list->moveStart",
  },
]
`;

exports[`mapActions stopMoveOutList 1`] = `
Array [
  Object {
    "type": "list->moveStop",
  },
]
`;

exports[`test mapState 1`] = `
Object {
  "dropArea": 2,
  "holder": 1,
  "orders": 3,
  "points": Array [],
}
`;

exports[`test mapState 2`] = `
Object {
  "dropArea": undefined,
  "holder": undefined,
  "orders": undefined,
  "points": 1,
}
`;
